src/object/cli_obj.c

```diff
@@ -2748,3 +2767,186 @@ dc_obj_query_key(tse_task_t *api_task)
+
+	sync_arg->sa_auxi.shard		= shard;
+	sync_arg->sa_auxi.target	= obj_shard2tgtid(obj, shard);
+	sync_arg->sa_auxi.map_ver	= map_ver;
+
+	return 0;
+}
+
+static int
+obj_shard_sync_task(tse_task_t *task)
+{
+	struct shard_sync_args	*args;
+	struct dc_object	*obj;
+	struct dc_obj_shard	*obj_shard;
+	int			 rc;
+
+	args = tse_task_buf_embedded(task, sizeof(*args));
+	obj = args->sa_auxi.obj;
+
+	rc = obj_shard_open(obj, args->sa_auxi.shard, args->sa_auxi.map_ver,
+			    &obj_shard);
+	if (rc != 0) {
+		if (rc == -DER_NONEXIST)
+			rc = 0;
+
+		tse_task_complete(task, rc);
+	} else {
+		rc = dc_obj_shard_sync(obj_shard, args->sa_coh_uuid,
+				       args->sa_cont_uuid, &args->sa_auxi.epoch,
+				       &args->sa_auxi.map_ver, task);
+
+		obj_shard_close(obj_shard);
+	}
+
+	return rc;
+}
+
+int
+dc_obj_sync(tse_task_t *api_task)
+{
+	struct daos_obj_sync	*api_args;
+	struct obj_auxi_args	*obj_auxi;
+	tse_sched_t		*sched = tse_task2sched(api_task);
+	struct dc_object	*obj = NULL;
+	d_list_t		*head = NULL;
+	daos_handle_t		 coh;
+	uuid_t			 coh_uuid;
+	uuid_t			 cont_uuid;
+	uint32_t		 replicas;
+	uint32_t		 map_ver;
+	int			 i = 0;
+	int			 rc;
+
+	if (srv_io_mode != DIM_DTX_FULL_ENABLED)
+		D_GOTO(out_task, rc = 0);
+
+	api_args = dc_task_get_args(api_task);
+	D_ASSERTF(api_args != NULL,
+		  "Task Argument OPC does not match DC OPC\n");
+
+	obj = obj_hdl2ptr(api_args->oh);
+	if (obj == NULL)
+		D_GOTO(out_task, rc = -DER_NO_HDL);
+
+	obj_auxi = tse_task_stack_push(api_task, sizeof(*obj_auxi));
+	obj_auxi->obj_task = api_task;
+	obj_auxi->opc = DAOS_OBJ_RPC_SYNC;
+	obj_auxi->epoch = api_args->epoch;
+	shard_task_list_init(obj_auxi);
+
+	rc = tse_task_register_comp_cb(api_task, obj_comp_cb, &obj,
+				       sizeof(obj));
+	if (rc != 0) {
+		/* NB: obj_comp_cb() will release refcount in other cases */
+		obj_decref(obj);
+		D_GOTO(out_task, rc);
+	}
+
+	coh = dc_obj_hdl2cont_hdl(api_args->oh);
+	if (daos_handle_is_inval(coh))
+		D_GOTO(out_task, rc = -DER_NO_HDL);
+
+	rc = dc_cont_hdl2uuid(coh, &coh_uuid, &cont_uuid);
+	if (rc != 0)
+		D_GOTO(out_task, rc);
+
+	rc = obj_ptr2pm_ver(obj, &map_ver);
+	if (rc)
+		D_GOTO(out_task, rc);
+
+	obj_auxi->map_ver_req = map_ver;
+	head = &obj_auxi->shard_task_head;
+
+	/* For retried case, reuse the previous shard tasks and resched it */
+	if (obj_auxi->io_retry && obj_auxi->args_initialized) {
+		struct shard_task_reset_sync_args	arg;
+
+		arg.obj		= obj;
+		arg.map_ver	= map_ver;
+		tse_task_list_traverse(head, shard_task_reset_sync, &arg);
+
+		goto task_sched;
+	}
+
+	D_ASSERT(!obj_auxi->args_initialized);
+	D_ASSERT(d_list_empty(head));
+
+	/* Need to mark sync epoch on server for barrier, so even if the
+	 * @replicas is 1, we still need to send SYNC RPC to the server.
+	 */
+	replicas = obj_get_replicas(obj);
+
+	/* The SYNC RPC will be sent to the leader in each redundancy group. */
+	for (i = 0; i < obj->cob_shards_nr; i += replicas) {
```

Comment:
do you think the obj_req_fanout() style is better or worse?
I changed obj_update/fetch etc to that style so need not write the tse code for each obj API. If you think it is fine, I can change this API to that style later after this patch merged.

Reply:
OK, I will unify it with obj_req_fanout().